lalrpop-util/src/lexer.rs

```diff
@@ -10,56 +16,57 @@ impl<'a> fmt::Display for Token<'a> {
     }
 }
 
-struct RegexEntry {
-    regex: regex::Regex,
-    skip: bool,
-}
-
 pub struct MatcherBuilder {
-    regex_set: regex::RegexSet,
-    regex_vec: Vec<RegexEntry>,
+    dfa: dense::DFA<Vec<u32>>,
+    skip_vec: Vec<bool>,
 }
 
 impl MatcherBuilder {
+    #[allow(clippy::result_large_err)]
     pub fn new<S>(
         exprs: impl IntoIterator<Item = (S, bool)>,
-    ) -> Result<MatcherBuilder, regex::Error>
+    ) -> Result<MatcherBuilder, dense::BuildError>
     where
         S: AsRef<str>,
     {
         let exprs = exprs.into_iter();
         let mut regex_vec = Vec::with_capacity(exprs.size_hint().0);
-        let mut first_error = None;
-        let regex_set_result = regex::RegexSet::new(exprs.scan((), |_, (s, skip)| {
-            regex_vec.push(match regex::Regex::new(s.as_ref()) {
-                Ok(regex) => RegexEntry { regex, skip },
-                Err(err) => {
-                    first_error = Some(err);
-                    return None;
-                }
-            });
-            Some(s)
-        }));
-
-        if let Some(err) = first_error {
-            return Err(err);
+        let mut skip_vec = Vec::with_capacity(exprs.size_hint().0);
+        for (regex, skip) in exprs {
+            regex_vec.push(regex);
+            skip_vec.push(skip);
         }
-        let regex_set = regex_set_result?;
 
-        Ok(MatcherBuilder {
-            regex_set,
-            regex_vec,
-        })
+        let enable_unicode = cfg!(feature = "unicode");
+        let dfa = dense::Builder::new()
+            .configure(
+                dense::Config::new()
+                    .minimize(true)
+                    .start_kind(StartKind::Anchored)
+                    .match_kind(MatchKind::All),
+            )
+            .syntax(
+                SyntaxConfig::new()
+                    .unicode(enable_unicode)
+                    .utf8(enable_unicode),
+            )
+            .thompson(NfaConfig::new().utf8(enable_unicode).shrink(true))
+            .build_many(&regex_vec)?;
+
+        Ok(MatcherBuilder { dfa, skip_vec })
     }
+
     pub fn matcher<'input, 'builder, E>(
         &'builder self,
-        s: &'input str,
+        text: &'input str,
     ) -> Matcher<'input, 'builder, E> {
+        let start = self.dfa.universal_start_state(Anchored::Yes).unwrap();
```

Comment:
It's guaranteed to return `Some` only if there is no lookarounds. So regexes starting with `^` will return `None`. I don't think it's checked anywhere?

https://docs.rs/regex-automata/0.3.4/regex_automata/dfa/trait.Automaton.html#method.universal_start_state

Reply:
Originally lalrpop prepend `^` to each regex and I have already removed that operation. I'm not sure if it's checked.

Reply:
If user regexes contain `^` or `\b`, it should not panic at runtime at least.

Reply:
Fortunately, I just found that it's already checked here https://github.com/QuarticCat/lalrpop/blob/5667ae493206e07a7f572d9702869d4c4bbe0422/lalrpop/src/normalize/token_check/mod.rs#L354.

Reply:
> Fortunately, I just found that it's already checked here

Fair enough. Maybe use `.expect` and put the rationale there.

Reply:
Addressed